src/state/pwallet/concentratedLiquidity/hooks/evm.ts

```diff
@@ -0,0 +1,87 @@
+import { BigNumber } from 'ethers';
+import { useMemo } from 'react';
+import { usePangolinWeb3 } from 'src/hooks';
+import { useV3NFTPositionManagerContract } from 'src/hooks/useContract';
+import { useSingleCallResult, useSingleContractMultipleData } from 'src/state/pmulticall/hooks';
+import { UseV3PositionsResults } from '../types';
+
+function useV3PositionsFromTokenIds(tokenIds: BigNumber[] | undefined): UseV3PositionsResults {
+  const positionManager = useV3NFTPositionManagerContract();
+  const inputs = useMemo(() => (tokenIds ? tokenIds.map((tokenId) => [BigNumber.from(tokenId)]) : []), [tokenIds]);
+  const results = useSingleContractMultipleData(positionManager, 'positions', inputs);
+
+  const loading = useMemo(() => results.some(({ loading }) => loading), [results]);
+  const error = useMemo(() => results.some(({ error }) => error), [results]);
+
+  const positions = useMemo(() => {
+    if (!loading && !error && tokenIds) {
+      return results.map((call, i) => {
+        const tokenId = tokenIds[i];
+        const result = call.result as any; // any => CallStateResult
+        return {
+          tokenId,
+          fee: result.fee,
+          feeGrowthInside0LastX128: result.feeGrowthInside0LastX128,
+          feeGrowthInside1LastX128: result.feeGrowthInside1LastX128,
+          liquidity: result.liquidity,
+          nonce: result.nonce,
+          operator: result.operator,
+          tickLower: result.tickLower,
+          tickUpper: result.tickUpper,
+          token0: result.token0,
+          token1: result.token1,
+          tokensOwed0: result.tokensOwed0,
+          tokensOwed1: result.tokensOwed1,
+        };
+      });
+    }
+    return undefined;
+  }, [loading, error, results, tokenIds]);
+
+  return {
+    loading,
+    positions: positions?.map((position, i) => ({ ...position, tokenId: inputs[i][0] })),
+  };
+}
+
+export function useGetUserPositions() {
```

Comment:
can you add some comments above the hook on what its doing ? also please add some comments above useV3PositionsFromTokenIds